File: backend/app/api/v1/asset_collections.py
from fastapi import APIRouter, Depends, Response, status

from app.api.dependencies import (
    get_asset_collection_service,
    get_current_user,
    require_csrf,
)
from app.models.user import User
from app.schemas.asset_collection import (
    AssetCollectionBoardCreate,
    AssetCollectionBoardListResponse,
    AssetCollectionBoardRead,
    AssetCollectionBoardUpdate,
    AssetCollectionMembership,
    AssetCollectionSummary,
    AssetSaveContext,
    SavedAssetListResponse,
)
from app.services.asset_collection_service import AssetCollectionService

router = APIRouter(prefix="/me/asset-collections", tags=["asset-collections"])


@router.get("/summary", response_model=AssetCollectionSummary)
def get_summary(
    user: User = Depends(get_current_user),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.summary(user)


@router.get("/assets/{asset_group_id}/membership", response_model=AssetCollectionMembership)
def get_membership(
    asset_group_id: str,
    user: User = Depends(get_current_user),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.membership(user, asset_group_id)


@router.get("/likes", response_model=SavedAssetListResponse)
def list_likes(
    cursor: str | None = None,
    limit: int = 48,
    user: User = Depends(get_current_user),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.list_likes(user, cursor=cursor, limit=max(1, min(limit, 100)))


@router.put("/likes/{asset_group_id}", response_model=AssetCollectionMembership)
def like_asset(
    asset_group_id: str,
    payload: AssetSaveContext,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.like(user, asset_group_id, payload)


@router.delete("/likes/{asset_group_id}", response_model=AssetCollectionMembership)
def unlike_asset(
    asset_group_id: str,
    payload: AssetSaveContext,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.unlike(user, asset_group_id, payload)


@router.get("/boards", response_model=AssetCollectionBoardListResponse)
def list_boards(
    user: User = Depends(get_current_user),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.list_boards(user)


@router.post(
    "/boards",
    response_model=AssetCollectionBoardRead,
    status_code=status.HTTP_201_CREATED,
)
def create_board(
    payload: AssetCollectionBoardCreate,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.create_board(user, payload.name)


@router.patch("/boards/{board_id}", response_model=AssetCollectionBoardRead)
def rename_board(
    board_id: str,
    payload: AssetCollectionBoardUpdate,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.rename_board(user, board_id, payload.name)


@router.delete("/boards/{board_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_board(
    board_id: str,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    service.delete_board(user, board_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/boards/{board_id}/items", response_model=SavedAssetListResponse)
def list_board_items(
    board_id: str,
    cursor: str | None = None,
    limit: int = 48,
    user: User = Depends(get_current_user),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.list_board_items(
        user, board_id, cursor=cursor, limit=max(1, min(limit, 100))
    )


@router.put(
    "/boards/{board_id}/items/{asset_group_id}",
    response_model=AssetCollectionMembership,
)
def add_to_board(
    board_id: str,
    asset_group_id: str,
    payload: AssetSaveContext,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.add_to_board(user, board_id, asset_group_id, payload)


@router.delete(
    "/boards/{board_id}/items/{asset_group_id}",
    response_model=AssetCollectionMembership,
)
def remove_from_board(
    board_id: str,
    asset_group_id: str,
    payload: AssetSaveContext,
    user: User = Depends(require_csrf),
    service: AssetCollectionService = Depends(get_asset_collection_service),
):
    return service.remove_from_board(user, board_id, asset_group_id, payload)

File: client/src/api/assetCollections.ts
import { api } from './client';

import type {
  AssetCollectionBoard,
  AssetCollectionMembership,
  AssetCollectionSummary,
  AssetSaveContext,
  SavedAssetListResponse,
} from '@client/src/types/api';

const ROOT = '/api/me/asset-collections';

export async function fetchAssetCollectionSummary(): Promise<AssetCollectionSummary> {
  return (await api.get(`${ROOT}/summary`)).data;
}

export async function fetchAssetMembership(
  assetGroupId: string,
): Promise<AssetCollectionMembership> {
  return (await api.get(`${ROOT}/assets/${assetGroupId}/membership`)).data;
}

export async function fetchLikedAssets(cursor?: string): Promise<SavedAssetListResponse> {
  return (await api.get(`${ROOT}/likes`, { params: { cursor, limit: 100 } })).data;
}

export async function likeAsset(
  assetGroupId: string,
  context: AssetSaveContext,
): Promise<AssetCollectionMembership> {
  return (await api.put(`${ROOT}/likes/${assetGroupId}`, context)).data;
}

export async function unlikeAsset(
  assetGroupId: string,
  context: AssetSaveContext,
): Promise<AssetCollectionMembership> {
  return (await api.delete(`${ROOT}/likes/${assetGroupId}`, { data: context })).data;
}

export async function fetchCollectionBoards(): Promise<AssetCollectionBoard[]> {
  return (await api.get(`${ROOT}/boards`)).data.items;
}

export async function createCollectionBoard(name: string): Promise<AssetCollectionBoard> {
  return (await api.post(`${ROOT}/boards`, { name })).data;
}

export async function renameCollectionBoard(
  boardId: string,
  name: string,
): Promise<AssetCollectionBoard> {
  return (await api.patch(`${ROOT}/boards/${boardId}`, { name })).data;
}

export async function deleteCollectionBoard(boardId: string): Promise<void> {
  await api.delete(`${ROOT}/boards/${boardId}`);
}

export async function fetchCollectionBoardItems(
  boardId: string,
  cursor?: string,
): Promise<SavedAssetListResponse> {
  return (await api.get(`${ROOT}/boards/${boardId}/items`, {
    params: { cursor, limit: 100 },
  })).data;
}

export async function addAssetToBoard(
  boardId: string,
  assetGroupId: string,
  context: AssetSaveContext,
): Promise<AssetCollectionMembership> {
  return (await api.put(`${ROOT}/boards/${boardId}/items/${assetGroupId}`, context)).data;
}

export async function removeAssetFromBoard(
  boardId: string,
  assetGroupId: string,
  context: AssetSaveContext,
): Promise<AssetCollectionMembership> {
  return (
    await api.delete(`${ROOT}/boards/${boardId}/items/${assetGroupId}`, { data: context })
  ).data;
}
